### PR TITLE
Dynamic pow

### LINKED ISSUE #291: Make PoW difficulty changeable
Since we intend on adding an authority which the service nodes can poll to get the current PoW difficulty, we need a way for the clients to be informed about which PoW difficulty they should be using

The plan for now is to follow the "wrong swarm" error code format, simply return a specific 4XX error for requests with invalid PoW along with the current correct difficulty which the client can then update locally and retry.

### PR DESCRIPTION
**Pull request by BeaudanBrown**
Clients now save their PoW difficulty (formerly NONCE_TRIALS in the storage) and can update it when informed to do so by service nodes
There's a bunch of simple renaming because I changed nonce_trials to difficulty and also cleaned up loki_rpc a bit

closes #291

### CODE REVIEW

File: libloki/proof-of-work.js [hunk 1]
```diff
@@ -74,9 +74,9 @@ const pow = {
       ).toArrayBuffer()
     );
 
-    const nonceTrials =
-      _nonceTrials || (development ? DEV_NONCE_TRIALS : PROD_NONCE_TRIALS);
-    const target = pow.calcTarget(ttl, payload.length, nonceTrials);
+    const difficulty =
+      _difficulty || (development ? DEV_DIFFICULTY : PROD_DIFFICULTY);
```
Comment (Mikunj) [inline libloki/proof-of-work.js]:
```
development ? DEV_DIFFICULTY : PROD_DIFFICULTY
```
Is there a need for this anymore?

Comment (BeaudanBrown) [reply libloki/proof-of-work.js]: I left it there to cover the (hopefully impossible) case where `_difficulty` is somehow null or undefined
But now that you mention it I could probably just collapse the constants into a single fallback difficulty

The idea is that if you somehow broke your difficulty you would just make a request with the fallback difficulty and then the snode would inform you of the correct difficulty

Comment (Mikunj) [reply libloki/proof-of-work.js]: That would be a better idea.

File: js/modules/loki_message_api.js [hunk 2]
```diff
@@ -92,7 +94,22 @@ class LokiMessageAPI {
...
+          this.snodeServerPort,
+          'store',
+          params
+        );
+
+        // Make sure we aren't doing too much PoW
+        const currentDifficulty = window.storage.get('PoWDifficulty', null);
+        const newDifficulty = result.difficulty;
+        if (
+          !Number.isNaN(newDifficulty) &&
```
Comment (sachaaaaa) [inline js/modules/loki_message_api.js]: Just wondering if `result.difficulty` is a string or a number?

Comment (BeaudanBrown) [reply js/modules/loki_message_api.js]: Number

File: js/modules/loki_message_api.js [hunk 3]
```diff
@@ -164,6 +186,12 @@ class LokiMessageAPI {
           await lokiSnodeAPI.updateSwarmNodes(params.pubKey, newSwarm);
           this.sendingSwarmNodes[params.timestamp] = newSwarm;
           return false;
+        } else if (e instanceof textsecure.WrongDifficultyError) {
+          const { newDifficulty } = e;
+          if (!Number.isNaN(newDifficulty)) {
+            window.storage.put('PoWDifficulty', newDifficulty);
+          }
+          throw e;
```
Comment (sachaaaaa) [inline js/modules/loki_message_api.js]: do you want to throw even after updating the difficulty?
Otherwise `return false;` in the `if` block?

Comment (BeaudanBrown) [reply js/modules/loki_message_api.js]: Ya I explicitly throw here to force a restart of the sendMessage logic to actually *use* the new PoW difficulty
```js
    try {
      results = await Promise.all(promises);
    } catch (e) {
      if (e instanceof textsecure.WrongDifficultyError) {
        // Force nonce recalculation
        this.sendMessage(pubKey, data, messageTimeStamp, ttl, options);
        return;
      }
      throw e;
    }
```
Not the most elegant solution but gets the job done, can also be extended to different errors in the future to do different things if we want

Comment (sachaaaaa) [reply js/modules/loki_message_api.js]: Seems reasonable!

**Pull request merged**